### PR TITLE
[ RNN ] Add unittest case & return_sequence

### PR DESCRIPTION
**Pull request by jijoongmoon**
This PR includes.
  . unittest of rnn layer
  . support of return sequence of rnn layer

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>

### CODE REVIEW

File: nntrainer/layers/rnn.cpp [hunk 1]
```diff
@@ -81,6 +85,21 @@ int RNNLayer::initialize(Manager &manager) {
   h_prev = Tensor(bias_dim);
   h_prev.setZero();
 
+  TensorDim d = input_dim[0];
+  d.width(unit);
+
+  // We do not need this if we reuse net_hidden[0]. But if we do, then the unit
+  // test will fail. Becuase it modifies the date during gradient calculation
+  // TODO : We could control with something like #define test to save memory
```
Comment (kparichay) [inline nntrainer/layers/rnn.cpp:93]: agreed, we can define a macro from meson based on unit_test to disable some optimizations for unittests.

File: nntrainer/layers/rnn.cpp [hunk 2]
```diff
@@ -117,7 +142,13 @@ void RNNLayer::forwarding(bool training) {
   Tensor &bias_h =
     weightAt(static_cast<int>(RNNParams::bias_h)).getVariableRef();
 
-  Tensor &hidden_ = net_hidden[0]->getVariableRef();
+  Tensor hidden_;
+  // if (!return_sequences) {
```
Comment (kparichay) [inline nntrainer/layers/rnn.cpp:146]: is this intended?

File: nntrainer/layers/rnn.cpp [hunk 3]
```diff
@@ -158,6 +189,19 @@ void RNNLayer::forwarding(bool training) {
...
+  if (!return_sequences) {
+    TensorDim d = hidden_.getDim();
+    for (unsigned int b = 0; b < input_dim[0].batch(); ++b) {
+      float *data = hidden_.getAddress(b * d.width() * d.height() +
+                                       (d.height() - 1) * d.width());
+      float *rdata = net_hidden[0]->getVariableRef().getAddress(b * d.width());
+      std::copy(data, data + d.width(), rdata);
+    }
+  } else {
+    std::copy(hidden_.getData(), hidden_.getData() + hidden_.length(),
```
Comment (kparichay) [inline nntrainer/layers/rnn.cpp:202]: i recommend to use `Tensor::copy()`

File: nntrainer/layers/rnn.cpp [hunk 4]
```diff
@@ -158,6 +189,19 @@ void RNNLayer::forwarding(bool training) {
...
       h_prev.getBatchSlice(b, 1).copy(hs);
   }
+
+  if (!return_sequences) {
+    TensorDim d = hidden_.getDim();
+    for (unsigned int b = 0; b < input_dim[0].batch(); ++b) {
+      float *data = hidden_.getAddress(b * d.width() * d.height() +
+                                       (d.height() - 1) * d.width());
+      float *rdata = net_hidden[0]->getVariableRef().getAddress(b * d.width());
+      std::copy(data, data + d.width(), rdata);
```
Comment (kparichay) [inline nntrainer/layers/rnn.cpp:199]: please try and use `Tensor::copy()`

File: nntrainer/layers/rnn.cpp [hunk 5]
```diff
@@ -186,8 +232,27 @@ void RNNLayer::calcGradient() {
...
+  if (!return_sequences) {
+    TensorDim d = derivative_.getDim();
+    for (unsigned int b = 0; b < input_dim[0].batch(); ++b) {
+      float *data = derivative_.getAddress(b * d.width() * d.height() +
+                                           (d.height() - 1) * d.width());
+      float *rdata = net_hidden[0]->getGradientRef().getAddress(b * d.width());
+      std::copy(rdata, rdata + d.width(), data);
+    }
+  } else {
+    std::copy(net_hidden[0]->getGradientRef().getData(),
```
Comment (kparichay) [inline nntrainer/layers/rnn.cpp:248]: same

File: nntrainer/layers/rnn.cpp [hunk 6]
```diff
@@ -186,8 +232,27 @@ void RNNLayer::calcGradient() {
...
+  Tensor hidden_;
+  derivative_ = hidden->getGradientRef();
+
+  if (!return_sequences) {
+    TensorDim d = derivative_.getDim();
+    for (unsigned int b = 0; b < input_dim[0].batch(); ++b) {
+      float *data = derivative_.getAddress(b * d.width() * d.height() +
+                                           (d.height() - 1) * d.width());
+      float *rdata = net_hidden[0]->getGradientRef().getAddress(b * d.width());
+      std::copy(rdata, rdata + d.width(), data);
```
Comment (kparichay) [inline nntrainer/layers/rnn.cpp:245]: same

File: test/input_gen/genModelTests.py [hunk 7]
```diff
@@ -428,3 +428,40 @@ def addition_test():
...
+            rnn_layer,
+            K.layers.Dense(1)
+        ],
+        optimizer=opt.SGD(learning_rate=0.1),
+        iteration=1,
+        input_shape=(1,1,1),
+        label_shape=(1,1),
+        is_onehot=False,
+        loss_fn_str="mse"
+    )
```
Comment (zhoonit) [inline test/input_gen/genModelTests.py:446]: Could you consider raising for more complicated scenario? shape of 1 and iteration 1 might not be sufficient.

**Review by zhoonit — APPROVED** [review]
LGTM

File: nntrainer/layers/rnn.cpp (hunk 2 above)
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:146]: my bad. removed.

File: nntrainer/layers/rnn.cpp (hunk 4 above)
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:199]: This isn't tensor copy.

File: nntrainer/layers/rnn.cpp (hunk 3 above)
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:202]: done.

File: nntrainer/layers/rnn.cpp (hunk 6 above)
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:245]: same.

File: nntrainer/layers/rnn.cpp (hunk 5 above)
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:248]: done.

File: nntrainer/layers/rnn.cpp (hunk 4 above)
Comment (kparichay) [reply nntrainer/layers/rnn.cpp:195]: consider creating tensor wrappers of data, and rdata and then using tensor::copy.
it is not recommended to use std::copy, but rather use Tensor::copy for memory inside tensor.

Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:195]: I will change it later pr..

File: test/input_gen/genModelTests.py (hunk 7 above)
Comment (jijoongmoon) [reply test/input_gen/genModelTests.py:446]: there will be more unittest cases followed PR.

Comment (zhoonit) [reply test/input_gen/genModelTests.py:446]: Awesome :)

File: nntrainer/layers/rnn.cpp (hunk 4 above)
Comment (kparichay) [reply nntrainer/layers/rnn.cpp:195]: alright.

**Review by kparichay — APPROVED** [review]
LGTM

**Pull request merged**